Add labels! and labels_sep_by! utility macros.

dashed opened the pull request:
Closes https://github.com/Stebalien/horrorshow-rs/issues/7

I use suggestion from https://github.com/Stebalien/horrorshow-rs/issues/7#issuecomment-223713500

------

I added `labels_sep_by!` because I wanted to be able to customize delimiters. I use it to generate the style attribute of the element.

-----

@Stebalien I couldn't get this to work `labels!("active" if false)` since:

> expr and stmt variables may only be followed by one of: => , ;

according to https://doc.rust-lang.org/book/macros.html

Any suggestions?

dashed commented:
I realized I forgot to create a branch. 😓

Stebalien commented:
Nice!

> expr and stmt variables may only be followed by one of: => , ;
> 
> according to https://doc.rust-lang.org/book/macros.html
> 
> Any suggestions?

:cry: I was thinking of patterns. Oh well. `=>` is clear enough.

Stebalien commented on src/macros.rs:
```diff
@@ -327,3 +327,95 @@ macro_rules! template {
+    ($sep:expr; $item:expr) => {{
+
+        use $crate::Template;
+        use $crate::RenderOnce;
```
Instead of importing and calling, just use the fully qualified syntax: `$crate::RenderOnce::render(&$item, tmpl)`

Otherwise, you run the risk of running into conflicting functions.

Stebalien commented on src/macros.rs:
```diff
@@ -327,3 +327,95 @@ macro_rules! template {
+            if $should_include {
+                $item.render_once(tmpl);
+            }
+        }).into_string().unwrap()
```
Just return the template. No need to turn it into a `String`

Stebalien commented on src/macros.rs:
```diff
@@ -327,3 +327,95 @@ macro_rules! template {
+}
+
+#[macro_export]
+macro_rules! labels_sep_by {
```
Some documentation and an example would be helpful.

Stebalien commented:
Just a few nits and this looks good to go.

dashed commented:
I amended with documentation and suggested refactor.

I also refactored the macros to add separator only if there is a previous label. This was an edge case I missed.

Stebalien commented on src/macros.rs:
````diff
@@ -327,3 +327,148 @@ macro_rules! template {
+///
+/// This useful in generating class attribute as follows:
+///
+/// ```norun
````
Why mark this norun?

dashed replied:
I copied the doc comments from another macro and reworked it. I wasn't sure what `norun` does. I can remove it if you like.

Stebalien replied:
That's because I've been lazy :sleeping:. I just pushed a commit that removes the norun directives (so you can see how to do this). (...I also ended up having to fix many of the examples which is why norun should be avoided...).

Stebalien commented on src/macros.rs:
```diff
@@ -327,3 +327,148 @@ macro_rules! template {
     () => {}
 }
+
+/// Utility macro for conditionally joining labels, such as `String` or `&str`,
```
That's a bit ambiguous. The labels are *unconditionally* joined but *conditionally* included. "Utility macro for joining a set of conditionally included labels, ..." is the best I could come up with but that still sounds a bit awkward.

Stebalien commented on src/macros.rs:
```diff
@@ -327,3 +327,148 @@ macro_rules! template {
+
+    (@inner_expand $has_before:expr; $sep:expr; $tmpl:ident $item:expr) => {
+        if $has_before {
+            $crate::RenderOnce::render_once($sep, $tmpl);
```
`&$sep` is more versatile as it will trigger deref coercions.

dashed replied:
Should I do the same for `$item`? i.e. `&$item`

Stebalien commented on src/macros.rs:
```diff
@@ -327,3 +327,148 @@ macro_rules! template {
+            }
+            $crate::RenderOnce::render_once($item, $tmpl);
+            labels_sep_by!(@inner_expand true; $sep; $tmpl $($tail)*);
+        } else {
```
This will generate a lot of code. It may optimize away but will likely hurt compile times (not like the rest of horrorshow helps here...). Maybe just call the macro once and pass `$should_include || $has_before`?

Stebalien commented on src/macros.rs:
```diff
@@ -327,3 +327,148 @@ macro_rules! template {
+            if $should_include {
+                $crate::RenderOnce::render_once($item, tmpl);
+                labels_sep_by!(@inner_expand true; $sep; tmpl $($tail)*);
+            } else {
```
Same here. You can just pass `$should_include` instead of the `true` and `false` constants.

dashed commented:
I refactored and I reworked the docs a bit; hope it's less awkward :)

dashed commented:
Removing `norun` breaks travis.

Stebalien commented:
See the updated docs (where I've removed norun). Basically, you have to wrap your example:

```
# #[macro_use]
# extern crate horrorshow;
# fn main() {
call_macro! {
}
# }
```

The `#` prefix means "don't display this line".

Stebalien commented:
The new docs are great! Much better than I would have written, I hate writing docs... I'll merge after you fix the travis failure.

dashed commented:
I fixed the doctests and squashed the commits.

------

I think we ironed out all the issues for this PR.
@Stebalien , thanks for the assistance in this PR! 👍

Stebalien commented:
Thank you! It was a pleasure working with you.
